Typed unstash improvements

johanandren opened the pull request:
Fixes #26362 

* Makes unhandled from unstashing a message handled the same way as regular unhandled
* Allows stopping in the middle of unstashing (one ? here though about what to do with stash contents)
* Allows `same` as the initial behavior to start from current actor behavior when unstashing

johanandren committed:
Handle unhandled while unstashing #26362

johanandren committed:
Handle stop while unstashing

johanandren committed:
unstashing with initial Behaviors.same to refer to current actor behavior

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala:
```diff
@@ -125,20 +125,47 @@ import akka.util.ConstantFun
+        if (Behavior.isAlive(actualNext))
+          interpretOne(Behavior.canonicalize(actualNext, b2, ctx)) // recursive
+        else
+          // fixme rest of stash to dead letter?
```
yes, I think that is the right thing to do

helena replied:
Create issue so it's not lost in the depths and mark #number? Unless it's happening in this PR eventually of course.

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala:
```diff
@@ -125,20 +125,47 @@ import akka.util.ConstantFun
+    if (Behavior.isAlive(actualInitialBehavior)) {
+      interpretOne(actualInitialBehavior)
+    } else {
+      // fixme rest of stash to dead letter?
```
yes

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala:
```diff
@@ -125,20 +125,47 @@ import akka.util.ConstantFun
+        val actualNext =
+          if (interpretResult == Behavior.same) b2
+          else if (Behavior.isUnhandled(interpretResult)) {
+            ctx.asScala.asInstanceOf[ActorContextAdapter[_]].onUnhandled(message)
```
casting to `ActorContextAdapter` could cause trouble for Behavior testkit. Make it an internal method on `scaladsl.ActorContext` instead?

johanandren replied:
Ah, clever, didn't think of that

helena replied:
Do we have to reflect with asaInstanceOf?

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala:
```diff
@@ -125,20 +125,47 @@ import akka.util.ConstantFun
+      if (Behavior.isUnhandled(started))
+        throw new IllegalArgumentException("Cannot unstash with unhandled as starting behavior")
+      else if (started == Behavior.same) {
+        ctx.asScala.asInstanceOf[ActorContextAdapter[T]].currentBehavior
```
similar concern as above for Behavior testkit

patriknw commented on akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/StashBuffer.scala:
```diff
@@ -115,6 +117,8 @@ object StashBuffer {
    * messages will not be processed by this call and have to be unstashed
    * in another call.
+   *
+   * The `behavior` passed to `unstash` must not be `unhandled`
```
corresponding in `javadsl`

patriknw reviewed (commented):
great to fix those things

johanandren committed:
Unstash into deadletter when stopping during unstash

helena commented on akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StashSpec.scala:
```diff
@@ -562,26 +562,105 @@ class UnstashingSpec extends ScalaTestWithActorTestKit("""
+
+        Behaviors.receiveMessage {
+          case "unstash" =>
+            stash.unstashAll(ctx, Behaviors.receiveMessage {
```
Just curious why we went with `Behaviors.receiveMessage` versus `Behaviors.receive` for `Actor.receive` and to not be explicit or conflict with message/event.

johanandren replied:
I think `receive` just followed along from the untyped actor API. `receiveMessage` was added much later just to not have to deal with the context on every message unless you want to (not sure why it isn't just an overload though, it's `Function1` vs. `Function2` so erasure shouldn't be a problem, maybe something with type inference) 

In general I think the signals will be used much much less often so it's probably not _that_ bad that it is a bit inconsistent.

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/adapter/ActorContextAdapter.scala:
```diff
@@ -110,6 +110,12 @@ import scala.concurrent.duration._
   }
+
+  /**
+   * Made accessible only to allow stash to deal with unhandled messages just like if they were interpreted by
```
```suggestion
   * Made accessible to allow stash to deal with unhandled messages as though they were interpreted by
   * the adapter itself, even though the unstashing occurs inside the behavior stack.
```

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/scaladsl/StashBuffer.scala:
```diff
@@ -94,6 +94,8 @@ object StashBuffer {
    * messages will not be processed by this call and have to be unstashed
    * in another call.
+   *
+   * The initial `behavior` passed to `unstashAll` must not be `unhandled`
```
Are we afraid of `.` at the end of a doc sentence that is not an explicit `@param` description?

johanandren replied:
Nah, just sloppy I think

johanandren committed:
More unhandled test coverage

johanandren committed:
Avoid expecting ActorContextAdapter for internals

currentBehavior and onUnhandled moved to scaladsl.ActorContext as internal methods
allowing support for touching them also in the behavior testkit

johanandren committed:
Scaladoc fixes

helena commented on akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/internal/StubbedActorContext.scala:
```diff
@@ -289,4 +291,19 @@ private[akka] final class FunctionRef[-T](override val path: ActorPath, send: (T
+    unhandled = msg :: unhandled
+
+  /**
+   * Messages that are marked as unhandled
```
Nit, let's use periods on doc sentences that are not `@param` and friends?

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala:
```diff
@@ -125,20 +126,56 @@ import akka.util.ConstantFun
+        if (Behavior.isAlive(actualNext))
+          interpretOne(Behavior.canonicalize(actualNext, b2, ctx)) // recursive
+        else {
+          unstashRestToDeadLetters(ctx, messages)
```
👍

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala:
```diff
@@ -125,20 +126,56 @@ import akka.util.ConstantFun
+    val started = Behavior.start(behavior, ctx)
+    val actualInitialBehavior =
+      if (Behavior.isUnhandled(started))
+        throw new IllegalArgumentException("Cannot unstash with unhandled as starting behavior")
```
Punctuation is our friend.

helena commented on akka-actor-typed/src/main/scala/akka/actor/typed/internal/StashBufferImpl.scala:
```diff
@@ -125,20 +126,56 @@ import akka.util.ConstantFun
+    val untypedDeadLetters = scalaCtx.system.deadLetters.toUntyped
+    messages.foreach(msg =>
+      scalaCtx.system.deadLetters ! DeadLetter(msg, untypedDeadLetters, ctx.asScala.self.toUntyped))
   }
```
👍

helena reviewed (approved):
LGTM

johanandren committed:
Some more periods

patriknw commented on akka-persistence-typed/src/main/scala/akka/persistence/typed/scaladsl/EventSourcedBehavior.scala:
```diff
@@ -103,15 +102,10 @@ object EventSourcedBehavior {
-        }
-      case c =>
-        throw new IllegalStateException(s"Cannot extract the lastSequenceNumber from context ${c.getClass.getName}")
+    extractConcreteBehavior(context.currentBehavior) match {
```
nice

patriknw reviewed (approved):
LGTM
